src/compiler/scala/tools/nsc/typechecker/Typers.scala

```diff
@@ -3485,16 +3482,14 @@ trait Typers extends Adaptations with Tags with TypersTracking with PatternTyper
     def needsInstantiation(tparams: List[Symbol], formals: List[Type], args: List[Tree]) = {
       def isLowerBounded(tparam: Symbol) = !tparam.info.lowerBound.typeSymbol.isBottomClass
 
-      exists2(formals, args) {
-        case (formal, Function(vparams, _)) =>
-          (vparams exists (_.tpt.isEmpty)) &&
-          vparams.lengthIs <= MaxFunctionArity &&
-          (formal baseType FunctionClass(vparams.length) match {
+      tparams.forall(isLowerBounded) && exists2(formals, args) {
+        case (formal, Function(vparams, _)) if vparams.exists(_.tpt.isEmpty) =>
+          val arity = vparams.length
+          arity <= MaxFunctionArity && (formal.baseType(FunctionClass(arity)) match {
             case TypeRef(_, _, formalargs) =>
-              ( exists2(formalargs, vparams)((formal, vparam) =>
-                        vparam.tpt.isEmpty && (tparams exists formal.contains))
-                && (tparams forall isLowerBounded)
-              )
+              exists2(formalargs, vparams) { (formal, vparam) =>
+                vparam.tpt.isEmpty && tparams.exists(formal.contains)
+              }
             case _ =>
               false
           })
```

Comment:
I was hoping to lose the extra paren. :)  More seriously, hoisting the lower bounds check is probably cheap but it's incurred for every apply with undet tparams. It short-circuits looking at value args and other checks. A win!